### PR TITLE
[BUGFIX] Include public fields in property_names

### PR DESCRIPTION
**Pull request by kenwade4**
Include public fields in property_names so that the correct data is returned

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)

### CODE REVIEW

File: great_expectations/types/__init__.py [hunk 1]
```diff
@@ -188,12 +188,14 @@ def property_names(
...
         # Gather private fields:
         # By Python convention, properties of non-trivial length, prefixed by underscore ("_") character, are private.
-        property_names: Set[str] = set(
+        private_fields: Set[str] = set(
             filter(
                 lambda name: len(name) > 1,
                 [key[1:] for key in self.keys() if key[0] == "_"],
             )
         )
+        public_fields: Set[str] = {key for key in self.keys() if key[0] != "_"}
```
Comment (alexsherstinsky) [inline great_expectations/types/__init__.py:197]: @kenwade4 Is the idea here that in some objects we do not have private fields only, but also public fields like `self.my_field = "hello"`?  If so, then I would like to propose that for the objects that extend `DictDot` we avoid this pattern and instead only have private fields and `@property` getter and setters.  In fact, putting a comment here into the present module saying that this is what subclasses must do would be extremely useful.  Might you be able to change this?  Thanks!

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@kenwade4 I added a comment -- happy to discuss.

File: great_expectations/types/__init__.py (hunk 1 above)
Comment (kenwade4) [reply great_expectations/types/__init__.py:197]: The property_names set would only ever contain items from `self.keys()` if they started with an underscore. Everything else in this function just iterates over that property_names set via different comprehensions.

For any of the many objects we have that are like the following

```
@dataclass
class Whatever(SerializableDictDot):
    first: str
    second: Dict[str, Any]
    third: bool
```

... their `self.keys()` contains the properties defined (i.e. ["first", "second", "third"])... none of those start with an underscore and thus the property_names set would be empty and an empty set would be returned.

The `to_dict` method creates its internal new_dict by iterating over the keys returned by self.property_names(), so all those dataclasses just have an empty dict returned

Comment (alexsherstinsky) [reply great_expectations/types/__init__.py:197]: @kenwade4 Let's discuss -- this is important to get right.  Thanks!

**Review by alexsherstinsky — APPROVED** [review]
LGTM -- as long as all tests in Azure pass.  Thank you!

**Pull request merged**